### PR TITLE
Added Save/load IQR state

### PR DESCRIPTION
**Pull request by vbhavank**
Added save and load state but need to fix display of query image.

### CODE REVIEW

File: python/smqtk/web/search_app/modules/iqr/iqr_search.py [hunk 1]
```diff
@@ -300,13 +281,16 @@ def iqr_session_info():
                         tuple(d.uuid() for d in iqrs.negative_descriptors),
 
                     # UUIDs of example positive descriptors
-                    "ex_pos": tuple(self._iqr_example_pos_descr[iqrs.uuid]),
-                    "ex_neg": (),  # No user negative examples supported yet
+                    "external_pos": tuple(self._iqr_example_pos_descr[iqrs.uuid]),
+                    "external_neg": tuple(self._iqr_example_pos_descr[iqrs.uuid]),
```
Comment (Purg) [inline python/smqtk/web/search_app/modules/iqr/iqr_search.py]: This should probably map to an empty tuple instead of positive UUIDs. Negative external queries are not supported, so I support it does not matter either way...

File: python/smqtk/iqr/iqr_session.py [hunk 2]
```diff
@@ -324,8 +361,109 @@ def reset(self):
...
+            print("self._query_uuid",self._query_uuid)
+            ext_neg_d = d_set_to_list(self.external_negative_descriptors)
+        print("external_pos2:",ext_pos_d)
+        z_buffer = io.BytesIO()
+        print("keys",ext_pos_d[self._query_uuid].type())
+        z = zipfile.ZipFile(z_buffer, 'w', self.STATE_ZIP_COMPRESSION)
+        z.writestr(self.STATE_ZIP_FILENAME, json.dumps({
+            'pos': pos_d,
+            'neg': neg_d,
+            'external_pos': [(self._query_uuid,ext_pos_d[self._query_uuid].type(),self._query_f.tolist())],
```
Comment (Purg) [inline python/smqtk/iqr/iqr_session.py]: Is this line maybe why you are not seeing the external positive queries reappear on state load? I'm not sure what those values being set are, but they certainly are not ``self.external_positive_descriptors``.

File: python/smqtk/iqr/iqr_session.py [hunk 3]
```diff
@@ -324,8 +361,109 @@ def reset(self):
...
+            # Read in raw descriptor data from the state, convert to descriptor
+            # element, then store in our descriptor sets.
+            for source, target in [(state['external_pos'],
+                                    self.external_positive_descriptors),
+                                   (state['external_neg'],
+                                    self.external_negative_descriptors),
+                                   (state['pos'], self.positive_descriptors),
+                                   (state['neg'], self.negative_descriptors)]:
+                for uid, type_str, vector_list in source:
+                    e = load_descriptor(uid, type_str, vector_list)
```
Comment (Purg) [inline python/smqtk/iqr/iqr_session.py]: A more error-robust approach to this loop would be to load all descriptors, collecting them in a local list, before adding them to / updating the ``target`` set. That way if any descriptor load produces an error (the assert in the ``load_descriptor`` function) then the ``target`` set is not modified. Currently, the ``target`` set can end up in a partially modified state if there is an error part-way through.

File: python/smqtk/iqr/iqr_session.py [hunk 4]
```diff
@@ -324,8 +361,109 @@ def reset(self):
...
+                    target.add(e)
+            #external_descriptors(positive=(self._query_f.tolist()))
```
Comment (Purg) [inline python/smqtk/iqr/iqr_session.py]: Remove commented out code.

File: python/smqtk/utils/pytorch_metrics.py [hunk 5]
```diff
@@ -14,12 +14,15 @@ class DIS_TYPE(Enum):
 
 
 def L2_dis(t1, t2, dim):
-    # if not isinstance(t1, torch.Tensor):
-    #     raise TypeError("{} has to be torch.Tensor!".format(t1))
-    # if not isinstance(t2, torch.Tensor):
+    #print(t2.dtype())
```
Comment (Purg) [inline python/smqtk/utils/pytorch_metrics.py]: Remove commented out lines.

File: python/smqtk/web/search_app/modules/iqr/iqr_search.py [hunk 6]
```diff
@@ -315,45 +299,124 @@ def iqr_session_state():
...
+            )
+                 working_data = {}                 
+                 sid_data_elems = self._iqr_example_data.get(iqrs, {})
+                 for uid, workingElem in six.iteritems(sid_data_elems):
+                    working_data[uid] = {
+                    'content_type': workingElem.content_type(),
+                    'bytes_base64':
+                        base64.b64encode(workingElem.get_bytes()),
+                }
+                 print("working_data",working_data)
```
Comment (Purg) [inline python/smqtk/web/search_app/modules/iqr/iqr_search.py]: Probably want to remove this print so the log is not bloated with huge base64 strings.

File: python/smqtk/web/search_app/modules/iqr/iqr_search.py [hunk 7]
```diff
@@ -315,45 +299,124 @@ def iqr_session_state():
...
                 mimetype='application/octet-stream',
                 as_attachment=True,
-                attachment_filename="%s.IqrState" % iqrs_uuid,
+ 
+                attachment_filename="%s.IqrState" % str(iqrs.uuid)
             )
+        @self.route('/set_iqr_state',methods=["PUT"])
+        @self._parent_app.module_login.login_required
+        def set_iqr_session_state():
+
```
Comment (Purg) [inline python/smqtk/web/search_app/modules/iqr/iqr_search.py]: Remove empty line.

File: python/smqtk/web/search_app/modules/iqr/iqr_search.py [hunk 8]
```diff
@@ -315,45 +299,124 @@ def iqr_session_state():
...
+        @self._parent_app.module_login.login_required
+        def set_iqr_session_state():
+     
+            """
+            Set the current state based on the given state file.
+            """
+            sid = self.get_current_iqr_session()
+            fid = flask.request.form.get('fid', None)
+            state_base64 = flask.request.form.get('state_base64', None)
+            #data_base64 = flask.request.form.get('base64', None)
```
Comment (Purg) [inline python/smqtk/web/search_app/modules/iqr/iqr_search.py]: Remove commented out lines.

File: python/smqtk/web/search_app/modules/iqr/iqr_search.py [hunk 9]
```diff
@@ -315,45 +299,124 @@ def iqr_session_state():
...
+            service_zip.writestr(IqrSession.STATE_ZIP_FILENAME,
+                                 json.dumps(state_dict))
+            service_zip.close()
+            service_zip_base64 = \
+                base64.b64encode(service_zip_buffer.getvalue())
+
+            with self.get_current_iqr_session() as iqrs:
+                 state_bytes = \
+                base64.urlsafe_b64decode(service_zip_base64)
+                 #z_wrapper_buffer=iqrs.set_state_bytes(open(upload_filepath,'rb').read(),self._descr_elem_factory)
```
Comment (Purg) [inline python/smqtk/web/search_app/modules/iqr/iqr_search.py]: Remove commented out line.

File: python/smqtk/web/search_app/modules/iqr/iqr_search.py [hunk 10]
```diff
@@ -315,45 +299,124 @@ def iqr_session_state():
...
+                                          IqrSession.STATE_ZIP_COMPRESSION)
+            service_zip.writestr(IqrSession.STATE_ZIP_FILENAME,
+                                 json.dumps(state_dict))
+            service_zip.close()
+            service_zip_base64 = \
+                base64.b64encode(service_zip_buffer.getvalue())
+
+            with self.get_current_iqr_session() as iqrs:
+                 state_bytes = \
+                base64.urlsafe_b64decode(service_zip_base64)
```
Comment (Purg) [inline python/smqtk/web/search_app/modules/iqr/iqr_search.py]: Can this just be on one line?

File: python/smqtk/web/search_app/modules/iqr/static/js/smqtk.data_view.js [hunk 11]
```diff
@@ -18,7 +18,7 @@ function DataView(container, rank, uid, probability, saliency_flag, is_example,
     this.uid = uid;
     this.probability = probability;
     this.saliency_flag = saliency_flag;
-    this.is_example = is_example === undefined ? false : is_example;
+    this.is_example =  is_example;
```
Comment (Purg) [inline python/smqtk/web/search_app/modules/iqr/static/js/smqtk.data_view.js]: Was this change required?

Comment (vbhavank) [reply python/smqtk/web/search_app/modules/iqr/static/js/smqtk.data_view.js]: No, was a result of something i wanted to try. But changed it to the previous version now

File: python/smqtk/iqr/iqr_session.py (hunk 3 above)
Comment (vbhavank) [reply python/smqtk/iqr/iqr_session.py]: Removed load_descriptor and squashed its operations into the for loop so the assertion error can be caught by a try block

File: python/smqtk/iqr/iqr_session.py [hunk 12]
```diff
@@ -459,11 +444,17 @@ def load_descriptor(_uid, _type_str, vec_list):
...
+                    try:
+                        if _e.has_vector():
+                             assert _e.vector().tolist() == vector_list, \
+                             "Found existing vector for UUID '%s' but vectors did " \
+                              "not match."
+                        else:
+                             _e.set_vector(vector_list)
+                             target.add(_e)
+                    except:
+                        print("Trying to set a vector that is Incomplete/Invalid")
```
Comment (Purg) [inline python/smqtk/iqr/iqr_session.py]: This is now eating *any* exception that could be raised within the block. This also does not address the previous issue where you were updating the state while attempting to load vectors. The use of the ``load_descriptor`` function was fine and didn't need to be squashed.

I was more thinking of the following, assuming the previous ``load_descriptor`` function is defined:
```python
...
to_add_ex_pos = set()
to_add_ex_neg = set()
to_add_pos = set()
to_add_neg = set()
for source, target in [(state['external_pos'],
                        to_add_ex_pos),
                       (state['external_neg'],
                        to_add_ex_neg),
                       (state['pos'], to_add_pos),
                       (state['neg'], to_add_neg)]:
    for uid, type_str, vector_list in source:
        e = load_descriptor(uid, type_str, vector_list)
        target.append(e)
self.external_positive_descriptors.update(to_add_ex_pos)
self.external_negative_descriptors.update(to_add_ex_neg)
self.positive_descriptors.update(to_add_pos)
self.negative_descriptors(to_add_neg)
...
```

Comment (vbhavank) [reply python/smqtk/iqr/iqr_session.py]: Thanks for the edit. Fixed!

**Comment by Purg** [comment]
I think this is OK for now. @vbhavank mentioned offline that the lingering issue will be investigated for a later PR.
